prometheus/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"knative.dev/eventing/pkg/adapter"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/source"
+
+	"knative.dev/eventing-contrib/prometheus/pkg/apis/sources/v1alpha1"
+)
+
+const (
+	apiChunkOfURL = `/api/v1/query?query=`
+)
+
+type envConfig struct {
+	adapter.EnvConfig
+
+	EventSource string `envconfig:"EVENT_SOURCE" required:"true"`
+	ServerURL   string `envconfig:"PROMETHEUS_SERVER_URL" required:"true"`
+	PromQL      string `envconfig:"PROMETHEUS_PROM_QL" required:"true"`
+}
+
+type prometheusAdapter struct {
+	source    string
+	ce        cloudevents.Client
+	reporter  source.StatsReporter
+	namespace string
+	logger    *zap.SugaredLogger
+	serverURL string
+	promQL    string
+}
+
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &envConfig{}
+}
+
+// NewAdapter creates an adapter to convert PromQL replies to CloudEvents
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client, reporter source.StatsReporter) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	a := &prometheusAdapter{
+		source:    env.EventSource,
+		ce:        ceClient,
+		reporter:  reporter,
+		logger:    logger,
+		namespace: env.Namespace,
+		serverURL: env.ServerURL,
+		promQL:    env.PromQL,
+	}
+
+	return a
+}
+
+func (a *prometheusAdapter) Start(stopCh <-chan struct{}) error {
+	return wait.PollUntil(5*time.Second, a.send, stopCh)
```

Comment:
I think `wait.Until` is better here. `PollUntil` stops when there is an error, meaning that the adapter will stop running when an event couldn't be sent, which is not what we want. Instead we should log the error (for now) and keep going.

Reply:
I was wondering about that, too, tbh. And the hardcoded interval here.

Reply:
In that vein, https://github.com/knative/eventing/blob/ec4dcbbd5c35079dcbab80a94d22000338dda55b/pkg/adapter/main.go#L43 shouldn't return anything.

Reply:
Maybe. I'm not so sure that all errors can be caught at initialization time. But this is something to consider.